src/storage/txn/singleton_flow_controller.rs

```diff
@@ -0,0 +1,1584 @@
+        }
+        self.records
+            .iter()
+            .max_by(|a, b| a.0.partial_cmp(&b.0).unwrap())
+            .unwrap()
+            .0
+    }
+
+    pub fn get_percentile_90(&mut self) -> T {
+        if self.records.is_empty() {
+            return FromPrimitive::from_u64(0).unwrap();
+        }
+        let mut v: Vec<_> = self.records.iter().collect();
+        v.sort_by(|a, b| a.0.partial_cmp(&b.0).unwrap());
+        v[((self.records.len() - 1) as f64 * 0.90) as usize].0
+    }
+
+    pub fn trend(&self) -> Trend {
+        if self.records.len() <= 1 {
+            return Trend::NoTrend;
+        }
+
+        // If the lastest record is too old, no trend
+        if self.records.back().unwrap().1.saturating_elapsed_secs() > STALE_DUR as f64 {
+            return Trend::NoTrend;
+        }
+
+        let (mut left, mut left_cnt) = (T::default(), 0);
+        let (mut right, mut right_cnt) = (T::default(), 0);
+
+        // The time span matters
+        if MIN_TIME_SPAN > 0 {
+            // If the records doesn't cover a enough time span, no trend
+            let time_span = self.records.front().unwrap().1.saturating_elapsed_secs()
+                - self.records.back().unwrap().1.saturating_elapsed_secs();
+            if time_span < MIN_TIME_SPAN as f64 {
+                return Trend::NoTrend;
+            }
+
+            // Split the record into left and right by the middle of time range
+            for (_, r) in self.records.iter().enumerate() {
+                let elapsed_secs = r.1.saturating_elapsed_secs();
+                if elapsed_secs > time_span / 2.0 {
+                    left += r.0;
+                    left_cnt += 1;
+                } else {
+                    right += r.0;
+                    right_cnt += 1;
+                }
+            }
+        } else {
+            let half = self.records.len() / 2;
+            for (i, r) in self.records.iter().enumerate() {
+                if i < half {
+                    left += r.0;
+                    left_cnt += 1;
+                } else {
+                    right += r.0;
+                    right_cnt += 1;
+                }
+            }
+        }
+
+        // Decide if there is a trend by the two averages.
+        // Adding 2 here is to give a tolerance
+        let (l_avg, r_avg) = (left.as_() / left_cnt as f64, right.as_() / right_cnt as f64);
+        if r_avg > l_avg + 2.0 {
+            return Trend::Increasing;
+        }
+        if l_avg > r_avg + 2.0 {
+            return Trend::Decreasing;
+        }
+
+        Trend::NoTrend
+    }
+}
+
+// CFFlowChecker records some statistics and states related to one CF.
+// These statistics fall into five categories:
+//   * memtable
+//   * L0 files
+//   * L0 production flow (flush flow)
+//   * L0 consumption flow (compaction read flow of L0)
+//   * pending compaction bytes
+// And all of them are collected from the hook of RocksDB's event listener.
+struct CFFlowChecker {
```

Comment:
It should be `CfFlowChecker`.

Reply:
singleton_flow_controller.rs is just the git mv from flow_controller.rs. That's the existing code. But I think you're right.